lib/membrane_wav/parser.ex

```diff
@@ -59,114 +59,101 @@ defmodule Membrane.WAV.Parser do
-    demand = {:input, @init_stage_size}
-
-    {{:ok, demand: demand}, state}
+  def handle_caps(:input, _format, _context, state) do
+    {:ok, state}
   end
 
   @impl true
-  def handle_demand(:output, size, :bytes, _context, %{stage: :data} = state) do
-    {{:ok, demand: {:input, size}}, state}
+  def handle_process_list(:input, buffers, _context, %{stage: :data} = state) do
+    {{:ok, buffer: {:output, buffers}}, state}
   end
 
-  def handle_demand(
-        :output,
-        buffers_count,
-        :buffers,
-        _context,
-        %{stage: :data, frames_per_buffer: frames, format: format} = state
-      ) do
-    demand_size = RawAudio.frames_to_bytes(frames, format) * buffers_count
-    {{:ok, demand: {:input, demand_size}}, state}
-  end
+  def handle_process_list(:input, buffers, _context, state) do
+    payload =
+      buffers
+      |> Enum.map(&Map.get(&1, :payload))
+      |> List.insert_at(0, state.unparsed_data)
+      |> IO.iodata_to_binary()
 
-  def handle_demand(:output, _size, _unit, _context, state) do
-    {:ok, state}
+    {actions, state} = parse_payload(payload, state)
+
+    {{:ok, actions}, state}
   end
 
-  @impl true
-  def handle_process(:input, buffer, _context, %{stage: :data} = state) do
-    {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+  defp parse_payload(payload, state, actions_acc \\ [])
+
+  defp parse_payload(payload, %{stage: :data} = state, actions_acc) do
+    actions =
+      [{:buffer, {:output, %Buffer{payload: payload}}} | actions_acc]
+      |> Enum.reverse()
+
+    state = %{state | unparsed_data: ""}
+    {actions, state}
   end
 
-  def handle_process(
-        :input,
-        %Buffer{payload: payload} = _buffer,
-        _context,
-        %{stage: :init} = state
-      ) do
+  defp parse_payload(payload, %{stage: :init} = state, actions_acc)
+       when byte_size(payload) >= @init_stage_size do
     <<
       "RIFF",
       _file_size::32-little,
       "WAVE",
       "fmt ",
       format_chunk_size::32-little,
-      format::16-little
+      format::16-little,
+      rest::binary
     >> = payload
 
     check_format(format, format_chunk_size)
 
-    demand = {:input, @format_stage_size}
     state = %{state | stage: :format}
 
-    {{:ok, demand: demand}, state}
+    parse_payload(rest, state, actions_acc)
   end
 
-  def handle_process(
-        :input,
-        %Buffer{payload: payload} = _buffer,
-        _context,
-        %{stage: :format} = state
-      ) do
+  defp parse_payload(payload, %{stage: :format} = state, actions_acc)
+       when byte_size(payload) >= @format_stage_size do
     <<
       channels::16-little,
       sample_rate::32-little,
       _data_transmission_rate::32,
       _block_alignment_unit::16,
       bits_per_sample::16-little,
-      next_chunk_type::4-bytes,
-      next_chunk_size::32-little
+      next_chunk_type::32-bits,
+      next_chunk_size::32-little,
+      rest::binary
     >> = payload
```

Comment:
by moving this to the function header we could avoid the guard

Reply:
But it would make the header unreadable, so I prefer to keep the guard, especially since it better express the case where this clause shouldn't match